### PR TITLE
turn `unwrap_or` into `unwrap_or_else` and vice versa

### LINKED ISSUE #12983: Code action to transform between `foo_or` and `foo_or_else`
The standard library has a bunch of functions with two variants such as `unwrap_or` and `unwrap_or_else`, `ok_or` and `ok_or_else`.
The code action would transform for example:
`a.unwrap_or(calculate(x))` -> `a.unwrap_or_else(|| calculate(x))`
`a.unwrap_or(calculate())` -> `a.unwrap_or_else(calculate)`
and back
`a.unwrap_or_else(|| 5)` -> `a.unwrap_or(5)`

same for the other functions with `or` and `or_else` variants.

clippy currently has lints that warn against these, and suggests fixes, but it would be nice to have a code action for it :)
[clippy::or_fun_call](https://rust-lang.github.io/rust-clippy//master/index.html#or_fun_call)
[clippy::unnecessary_lazy_evaluations](https://rust-lang.github.io/rust-clippy//master/index.html#unnecessary_lazy_evaluations)

### PR DESCRIPTION
**Pull request by Austaras**
closes #12983

### CODE REVIEW

**Comment by A-Walrus** [comment]
This looks really good, thanks for working on it! :)
A few nits:

- Results don't have the `ok_or` /  `ok_or_else` methods.
- It would be nice to add methods for these (on both options and results) 
  - `or(5)` <-> `or_else(|| 5)` 
  -  `map_or(5,foo)` <-> `map_or_else(|| 5, foo)`

File: crates/ide-assists/src/handlers/replace_or_with_or_else.rs [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        [first, second] if map_or => {
+            let param = into_closure(first);
+            make::arg_list(vec![param, second.clone()])
+        }
+        _ => return None,
+    };
+
+    acc.add(
+        AssistId("replace_or_with_or_else", AssistKind::RefactorRewrite),
+        "Replace unwrap_or or ok_or with lazy version",
```
Comment (DorianListens) [inline crates/ide-assists/src/handlers/replace_or_with_or_else.rs:65]: Would it be worth making this title dynamic based on the context? Something like `format!(“Replace {name} with {replace}”)` maybe?

Comment (Austaras) [reply crates/ide-assists/src/handlers/replace_or_with_or_else.rs:65]: I prefer it to be aligned with function name

Comment (flodiebold) [reply crates/ide-assists/src/handlers/replace_or_with_or_else.rs:65]: We should be clear about what is being replaced by what here, so I agree with Dorian's suggestion.

Comment (Austaras) [reply crates/ide-assists/src/handlers/replace_or_with_or_else.rs:65]: It's not possible -- `AssistId` requires a static str

**Comment by jonas-schievink** [comment]
Thanks, this looks pretty good!

@bors r+

**Comment by lnicola** [comment]
changelog feature (first contribution) add assist to change between `XXX_or` and `XXX_or_else`

**Comment by lnicola** [comment]
https://user-images.githubusercontent.com/308347/188392529-be982a36-756a-400c-836d-b077de270c99.mp4

**Pull request merged**